Sources/Commands/PinsStore.swift

```diff
@@ -0,0 +1,188 @@
+        public let reason: String?
+
+        init(repository: RepositorySpecifier, version: Version, reason: String? = nil) {
+            self.repository = repository
+            self.version = version
+            self.reason = reason
+        }
+    }
+
+    /// The path to the pins file.
+    fileprivate let pinsFile: AbsolutePath
+
+    /// The filesystem to manage the pin file on.
+    fileprivate var fileSystem: FileSystem
+
+    /// The pins map.
+    fileprivate(set) var pinsMap: [RepositorySpecifier: Pin]
+
+    /// The current pins.
+    public var pins: AnySequence<Pin> {
+        return AnySequence<Pin>(pinsMap.values)
+    }
+
+    /// Create a new pins store.
+    ///
+    /// - Parameters:
+    ///   - pinsFile: Path to the pins file.
+    ///   - fileSystem: The filesystem to manage the pin file on.
+    public init(pinsFile: AbsolutePath, fileSystem: FileSystem) throws {
+        self.pinsFile = pinsFile
+        self.fileSystem = fileSystem
+        pinsMap = [:]
+        try restoreState()
+    }
+
+    /// Pin a repository at a version.
+    ///
+    /// - Parameters:
+    ///   - repository: The repository to pin. It should not be already pinned.
+    ///   - at: The version to pin at.
+    ///   - reason: The optional reason for pinning.
+    /// - Throws: PinOperationError
+    public mutating func pin(repository: RepositorySpecifier, at version: Version, reason: String? = nil) throws {
+        // The repo should not already be pinned.
+        guard pinsMap[repository] == nil else { throw PinOperationError.alreadyPinned }
+        // Add pin and save the state.
+        pinsMap[repository] = Pin(repository: repository, version: version, reason: reason)
+        try saveState()
+    }
+
+    /// Unpin a pinnned repository.
+    ///
+    /// - Parameters:
+    ///   - repository: The repository to unpin. It should already be pinned.
+    /// - Returns: The pin which was removed.
+    /// - Throws: PinOperationError
+    @discardableResult
+    public mutating func unpin(repository: RepositorySpecifier) throws -> Pin {
+        // The repo should already be pinned.
+        guard let pin = pinsMap[repository] else { throw PinOperationError.notPinned }
+        // Remove pin and save the state.
+        pinsMap[repository] = nil
+        try saveState()
+        return pin
+    }
+}
+
+/// Persistence.
+extension PinsStore {
+    // FIXME: A lot of the persistence mechanism here is copied from
+    // `RepositoryManager`. It would be nice to get actual infrastructure around
+    // persistence to handle the boilerplate parts.
+
+    private enum PersistenceError: Swift.Error {
+        /// The schema does not match the current version.
+        case invalidVersion
+
+        /// There was a missing or malformed key.
+        case unexpectedData
+    }
+
+    /// The current schema version for the persisted information.
+    ///
+    /// We currently discard any restored state if we detect a schema change.
```

Comment:
We can't do this, for a user managed data file. We will need to explicitly migrate when we change schemas...